docs/architecture/adr-0001-app-setup-and-goals.md

```diff
@@ -0,0 +1,168 @@
+
+`Model`s will be renamed to something less overloaded for developer sanity at
+some later date.
+
+### Notifying the API
+
+Since `model`s will be run in some asynchronous workflow, they'll need a way to
+notify the API. The form that notification should take is a [STAC item with the labeling extension](https://github.com/radiantearth/stac-spec/tree/master/extensions/label).
+Reasons to lean on STAC here are similar here to the reasons for leaning on it for new imagery. The API will accept a `POST` to a particular model
+that contains a STAC item with an asset that points to predictions on the new im
+agery.
+
+It's possible that future work could allow discrimination between predictions th
+at need time to run an asynchronous process and models that can do synchronous
+prediction. Since we don't know this is necessary or even possible, we'll assume
+for now that all predictions happen asynchronously and that all updates to the
+Granary API will happen via a later `POST` request.
+
+### Notifying end users
+
+When we create `model`s, we will also create AWS Simple Notification Service
+topics for those models. For our purposes we can think of these as an
+inexhaustible resource, since each AWS account is allowed 100,000 topics. At a
+time in the future when we think we are in danger of going over that limit we
+can think about a new strategy or ask for a service limit increase.
+Whenever a model is updated -- either a new prediction run has been kicked off,
+results are available, or metadata about that model has changed -- we'll send
+an event to the SNS topic for that model. The application will be deployable
+in any AWS account (including a client's account), so we assume that clients
+also have the ability to subscribe to those SNS topics and can subscribe
+to notifications.
+
+### Impact of continuous deployment
+
+There are several consequences of the choice to aim for continuous deployment:
+
+- Continuous integration checks passing on a PR _must_ mean that the branch is
+  safe to deploy. This will require more care around migrations than usual and
+  a more thorough exercise of the application in tests than we have in Raster
+  Foundry. As long as this repository is just an API, that probably means
+  generative testing of workflows modeled through API interaction both for
+  correctness and performance regressions. I believe that we already have the
+  tools and expertise on the team to do this.
+- This repository probably shouldn't follow the `git flow` pattern, since the
+  relationship between `develop` and `master` will be pretty
+  confused (`develop = master` at all times). It is currently `develop` for
+  consistency and to avoid the appearance of having made that choice final before
+  this ADR has been reviewed.
+- We'll need to determine a good story about feature flagging and probably also
+  vertically slice user-facing development. Alternatively, if vertical slicing
```

Comment:
The feature flagging component of this point makes sense to me, but the vertical slicing one doesn't. If the UI and API are decoupled by design, encoding a vertical slicing workflow to minimize friction between changes that span the frontend and backend seems a little brittle. As an alternative, I'm curious if things like a more robust frontend feature flagging framework, or API versioning were considered.

Reply:
API documentation with swagger becomes a bit tricky, since we'd have to maintain a lot of copies if we want the root spec document to be standalone... but it's possible we can lean on `$ref` and try to keep separate parts of the API distinctly documented to keep the API docs DRY

Reply:
Re the versioning strategy above, it at least does work:

```diff
diff --git a/app-backend/api/src/main/scala/com/rasterfoundry/granary/Server.scala b/app-backend/api/src/main/scala/com/rasterfoundry/granary/Server.scala
index c55a48b..1723565 100644
--- a/app-backend/api/src/main/scala/com/rasterfoundry/granary/Server.scala
+++ b/app-backend/api/src/main/scala/com/rasterfoundry/granary/Server.scala
@@ -10,15 +10,20 @@ import org.http4s.server.Router
 
 object ApiServer extends IOApp {
 
-  val httpApp: HttpApp[IO] = CORS(
+  val httpApp: HttpRoutes[IO] = CORS(
     Router(
-      "/api/hello"        -> HelloService.routes
-    )).orNotFound
+      "/api/hello" -> HelloService.routes
+    ))
+
+  val versioned: HttpApp[IO] = Router(
+    "/v1" -> httpApp,
+    "/v2" -> httpApp
+  ).orNotFound
 
   def run(args: List[String]): IO[ExitCode] =
     BlazeServerBuilder[IO]
       .bindHttp(8080, "0.0.0.0")
-      .withHttpApp(httpApp)
+      .withHttpApp(versioned)
       .serve
       .compile
       .drain
```

^^ that compiles

Reply:
I agree with Hector - probably not good to consider an approach without its own ADR - especially as versioning goes.

I think decoupling the frontend and backend development can be done without too much overhead and reliance on feature flags/coordination -- especially at an early stage for project development. If changes are additive we are usually fine doing the backend work first I think. In the case of breaking changes we will have to be more careful -- and that'll deserve its own ADR at that time.

Reply:
:+1: 96d5cc3, and also #2 and #3